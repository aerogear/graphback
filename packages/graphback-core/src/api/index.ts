
export * from "./ContextCreator"
export * from "./ContextTypes"
export * from "./GraphbackCRUDGeneratorConfig"
export * from "./GraphbackOperationType"