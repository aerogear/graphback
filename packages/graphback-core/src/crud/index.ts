export * from "./mappingHelpers"
export * from "./GraphbackOperationType"
