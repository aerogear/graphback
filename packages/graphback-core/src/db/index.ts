export * from './defaultNameTransforms';
export * from './modelTableMapper';
export * from './dataMapper';
export * from './getPrimaryKey';
