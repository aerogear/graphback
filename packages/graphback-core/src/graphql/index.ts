
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"
export * from "./directives"