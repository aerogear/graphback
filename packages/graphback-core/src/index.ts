
//Top level API
export * from './crud'
export * from './plugin/GraphbackPlugin'
export * from './plugin/GraphbackPluginEngine'
export * from './plugin/GraphbackGlobalConfig'
export * from './plugin/GraphbackCRUDGeneratorConfig';
export * from './plugin/ModelDefinition'
export * from './plugin/getSelectedFieldsFromResolverInfo';
export * from './plugin/GraphbackCoreMetadata'

export * from './relationships/RelationshipMetadataBuilder';
export * from './relationships/relationshipHelpers';
export * from './annotations/DefaultValueAnnotation';

export * from './utils/printSchemaWithDirectives';
export * from './utils/metadataAnnotations';
export * from './utils/fieldTransformHelpers';
export * from './utils/isTransientField';

export * from './runtime';
export * from './db';

export * from './scalars'
