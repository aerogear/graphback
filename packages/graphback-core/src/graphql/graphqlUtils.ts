import { readFileSync } from 'fs';
import { sync } from 'glob';
import { join } from 'path';
import { GraphbackOperationType, InputModelTypeContext, INTERFACE_TYPE_DEFINITION, OBJECT_TYPE_DEFINITION, OBJECT_TYPE_EXTENSION } from '../api';

export const getFieldName = (typeName: string, action: GraphbackOperationType, plural: string = ''): string => {
  const upperCasedType = typeName.charAt(0).toUpperCase() + typeName.substr(1);

  return `${action}${upperCasedType}${plural}`
}

export const getTableName = (typeName: string): string => {
  return typeName.toLowerCase()
}

export const filterObjectTypes = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === OBJECT_TYPE_DEFINITION);

export const filterInterfaceTypes = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === INTERFACE_TYPE_DEFINITION);

export const filterObjectExtensions = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === OBJECT_TYPE_EXTENSION);


