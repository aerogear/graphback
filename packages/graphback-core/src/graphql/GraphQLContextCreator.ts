import { parse, visit } from 'graphql';
import { InputModelTypeContext } from '../api/ContextTypes'
import { GraphbackCRUDGeneratorConfig } from "../api/GraphbackCRUDGeneratorConfig";
import { applyGeneratorDirectives } from './directives'
import { filterInterfaceTypes, filterObjectExtensions, filterObjectTypes } from './graphqlUtils';

import { InputContextCreator } from '../api';
import { inputTypeVisitor } from './InputTypeVisitor';



/**
 * Create input context to be reused for:
 * - schema
 * - various generators
 * - database creation
 */
export const graphQLInputContext: InputContextCreator = {

  createModelContext(schemaText: string, defaultConfig: GraphbackCRUDGeneratorConfig): InputModelTypeContext[] {
    const schema = applyGeneratorDirectives(schemaText)

    const astNode = parse(schema)
    const schemaDef = visit(astNode, { leave: inputTypeVisitor });
    const context = applyDefaultConfig(schemaDef, defaultConfig)
    const interfaces = filterInterfaceTypes(context)
    const finalTypes = applyExtensionsToOriginalTypes(context)

    return [...finalTypes, ...interfaces]

  }
}

// tslint:disable-next-line: no-any
function applyExtensionsToOriginalTypes(context: any) {
  const extendNodes = filterObjectExtensions(context)
  const filteredContext = filterObjectTypes(context)

  return filteredContext.map((t: InputModelTypeContext) => {
    const extendNode = extendNodes.find((node: InputModelTypeContext) => node.name === t.name);
    if (extendNode) {
      return {
        ...t,
        fields: [...t.fields, ...extendNode.fields],
        config: { ...t.config, ...extendNode.config },
        interfaces: [...t.interfaces, ...extendNode.interfaces]
      };
    }
    else {
      return t;
    }
  });
}

// tslint:disable-next-line: no-any
function applyDefaultConfig(schemaDef: any, defaultConfig: GraphbackCRUDGeneratorConfig) {
  return schemaDef.definitions.map((t: InputModelTypeContext) => {
    return {
      ...t,
      config: { ...defaultConfig, ...t.config }
    };
  });
}

