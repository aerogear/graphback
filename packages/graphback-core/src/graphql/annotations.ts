import { FieldDefinitionNode, TypeDefinitionNode } from 'graphql';
import { parseAnnotations } from 'graphql-annotations';
import { InputModelFieldAnnotations } from '../api';

export const parseFieldAnnotations = (node: FieldDefinitionNode): InputModelFieldAnnotations => {
    const fieldAnnotations: InputModelFieldAnnotations = {};

    const dbAnnotations: any = node.description ? parseAnnotations('db', String(node.description)) : undefined;

    if (!dbAnnotations) {
        return fieldAnnotations;
    }

    if (dbAnnotations.oneToOne) {
        fieldAnnotations.OneToOne = {
            field: dbAnnotations.oneToOne
        }
    }
    if (dbAnnotations.oneToMany) {
        fieldAnnotations.OneToMany = {
            field: dbAnnotations.oneToMany
        }
    }
    if (dbAnnotations.manyToOne) {
        fieldAnnotations.ManyToOne = {
            field: dbAnnotations.manyToOne
        }
    }
    if (dbAnnotations.manyToMany) {
        fieldAnnotations.ManyToMany = {
            field: dbAnnotations.manyToMany
        }
    }

    return fieldAnnotations;
}


export const parseTypeAnnotations = (node: TypeDefinitionNode) => {

    return node.description ? parseAnnotations('crud', String(node.description)) : undefined;
}
