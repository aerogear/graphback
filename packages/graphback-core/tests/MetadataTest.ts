import { buildSchema } from 'graphql';
import { GraphbackCoreMetadata, GraphbackCRUDGeneratorConfig, getModelByName } from '../src'

const setup = (model: string, config?: { crudMethods?: GraphbackCRUDGeneratorConfig }) => {
  const metadata = new GraphbackCoreMetadata({ crudMethods: config?.crudMethods }, buildSchema(model))

  return { metadata }
}

test('Model has default crud configuration', async () => {

  const { metadata } = setup(`

"""
@model
"""
type Comment {
  id: ID!
}

"""
@model
@delta
"""
type Note {
  id: ID!
  title: String!
  """
  @oneToMany field: 'note'
  """
  comments: [Comment]
}`)


  const models = metadata.getModelDefinitions();

  const { crudOptions, config, relationships } = getModelByName('Note', models)

  expect(crudOptions).toBeDefined()
  expect(config.deltaSync).toBe(true)
  expect(relationships).toHaveLength(1)
});

test('Model has default crud configuration', async () => {

  const { metadata } = setup(`
"""
@model
"""
type Note {
  id: ID!
  title: String!
}`)


  const models = metadata.getModelDefinitions();

  const { crudOptions } = getModelByName('Note', models)

  expect(crudOptions).toEqual({
    findOne: true,
    find: true,
    create: true,
    update: true,
    delete: true,
    subCreate: true,
    subDelete: true,
    subUpdate: true
  })
});

test('Override CRUD config for model', async () => {

  const { metadata } = setup(`
"""
@model(create: false)
"""
type Note {
  id: ID!
  title: String!
}

"""
@model
"""
type Comment {
  id: ID!
  title: String!
}
`)

  const models = metadata.getModelDefinitions();

  const note = getModelByName('Note', models)
  const comment = getModelByName('Comment', models)

  expect(note.crudOptions.create).toBe(false)
  expect(comment.crudOptions.create).toBe(true)
});

