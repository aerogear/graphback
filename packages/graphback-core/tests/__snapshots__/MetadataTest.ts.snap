// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test metadata 1`] = `
Array [
  Object {
    "create": true,
    "delete": false,
    "find": false,
    "findAll": true,
    "subCreate": true,
    "subDelete": true,
    "subUpdate": true,
    "update": true,
  },
  Object {
    "create": false,
    "delete": true,
    "find": false,
    "findAll": true,
    "subCreate": true,
    "subDelete": true,
    "subUpdate": true,
    "update": true,
  },
]
`;
